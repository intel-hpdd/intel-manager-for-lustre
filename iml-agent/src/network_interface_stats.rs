use combine::{
    error::ParseError,
    many1,
    parser::char::{char, digit, letter, spaces},
    stream::Stream,
    token, Parser,
};
use iml_wire_types::{InterfaceStats, RxStats, StatResult, TxStats};
use std::{collections::HashMap, convert::TryFrom, num::ParseIntError};

pub type InterfaceStatsMap = HashMap<String, InterfaceStats>;

fn interface<I>() -> impl Parser<I, Output = String>
where
    I: Stream<Token = char>,
    I::Error: ParseError<I::Token, I::Range, I::Position>,
{
    many1(letter().or(digit()).or(token('@')))
}

fn parse_interface<I>() -> impl Parser<I, Output = String>
where
    I: Stream<Token = char>,
    I::Error: ParseError<I::Token, I::Range, I::Position>,
{
    spaces()
        .and(many1(interface()))
        .map(|x| x.1)
        .skip(char(':'))
}

fn parse_stat<I>() -> impl Parser<I, Output = StatResult>
where
    I: Stream<Token = char>,
    I::Error: ParseError<I::Token, I::Range, I::Position>,
{
    spaces()
        .and(many1(digit()).map(|x: String| x.parse::<u64>()))
        .map(|x| x.1)
}

fn parse_stats<I>() -> impl Parser<
    I,
    Output = (
        StatResult,
        StatResult,
        StatResult,
        StatResult,
        StatResult,
        StatResult,
        StatResult,
        StatResult,
    ),
>
where
    I: Stream<Token = char>,
    I::Error: ParseError<I::Token, I::Range, I::Position>,
{
    spaces()
        .and(parse_stat())
        .and(parse_stat())
        .and(parse_stat())
        .and(parse_stat())
        .and(parse_stat())
        .and(parse_stat())
        .and(parse_stat())
        .and(parse_stat())
        .map(|((((((((_, x1), x2), x3), x4), x5), x6), x7), x8)| (x1, x2, x3, x4, x5, x6, x7, x8))
}

fn parse_stats_line<I>() -> impl Parser<I, Output = (String, Result<InterfaceStats, ParseIntError>)>
where
    I: Stream<Token = char>,
    I::Error: ParseError<I::Token, I::Range, I::Position>,
{
    parse_interface()
        .and(parse_stats())
        .and(parse_stats())
        .map(|((interface, rx), tx)| {
            let rx_stats = RxStats::try_from(rx);
            let tx_stats = TxStats::try_from(tx);

            (interface, InterfaceStats::try_from((rx_stats, tx_stats)))
        })
}

pub fn parse(output: &str) -> Result<InterfaceStatsMap, ParseIntError> {
    output
        .split('\n')
        .skip(2)
        .filter_map(|x| parse_stats_line().parse(x).ok())
        .try_fold(
            InterfaceStatsMap::new(),
            |mut acc, ((interface, stats), _)| match stats {
                Ok(stats) => {
                    let parts: Vec<&str> = interface.split('@').collect();
                    acc.insert(parts[0].to_string(), stats);

                    Ok(acc)
                }
                Err(e) => Err(e),
            },
        )
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_parse_interface() {
        let result = parse_stats_line()
            .parse("eth0: 72453387   55109    0    0    0     0          0         0   630390    9575    0    0    0     0       0          0")
            .map(|((interface, _), _)| interface);

        assert_eq!(result, Ok("eth0".to_string()));
    }

    #[test]
    fn test_parse_interface_stats() {
        let result = parse_stats_line()
            .parse("eth0: 72453387   55109    0    0    0     0          0         0   630390    9575    0    0    0     0       0          0")
            .map(|((_, stats), _)| stats);

        assert_eq!(
            result,
            Ok(Ok(InterfaceStats {
                rx: RxStats {
                    bytes: 72453387,
                    packets: 55109,
                    errs: 0,
                    drop: 0,
                    fifo: 0,
                    frame: 0,
                    compressed: 0,
                    multicast: 0
                },
                tx: TxStats {
                    bytes: 630390,
                    packets: 9575,
                    errs: 0,
                    drop: 0,
                    fifo: 0,
                    colls: 0,
                    carrier: 0,
                    compressed: 0
                }
            }))
        );
    }

    #[test]
    fn test_parse() {
        let stats_data = include_bytes!("./fixtures/network_stats.txt");
        let stats_data = std::str::from_utf8(stats_data).unwrap();
        let stats = parse(stats_data).unwrap();

        insta::with_settings!({sort_maps => true}, {
            insta::assert_json_snapshot!(stats)
        });
    }
}
