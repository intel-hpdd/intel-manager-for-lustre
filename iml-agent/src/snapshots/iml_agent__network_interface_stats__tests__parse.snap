---
source: iml-agent/src/network_interface_stats.rs
expression: stats
---
{
  "docker0": {
    "rx": {
      "bytes": 0,
      "packets": 0,
      "errs": 0,
      "drop": 0,
      "fifo": 0,
      "frame": 0,
      "compressed": 0,
      "multicast": 0
    },
    "tx": {
      "bytes": 0,
      "packets": 0,
      "errs": 0,
      "drop": 0,
      "fifo": 0,
      "colls": 0,
      "carrier": 0,
      "compressed": 0
    }
  },
  "ib0": {
    "rx": {
      "bytes": 4828965068,
      "packets": 23837574,
      "errs": 0,
      "drop": 0,
      "fifo": 0,
      "frame": 0,
      "compressed": 0,
      "multicast": 0
    },
    "tx": {
      "bytes": 4209864628,
      "packets": 26019919,
      "errs": 0,
      "drop": 0,
      "fifo": 0,
      "colls": 0,
      "carrier": 0,
      "compressed": 0
    }
  },
  "ib1": {
    "rx": {
      "bytes": 157149372,
      "packets": 1323290,
      "errs": 0,
      "drop": 0,
      "fifo": 0,
      "frame": 0,
      "compressed": 0,
      "multicast": 0
    },
    "tx": {
      "bytes": 5955084,
      "packets": 99249,
      "errs": 0,
      "drop": 0,
      "fifo": 0,
      "colls": 0,
      "carrier": 0,
      "compressed": 0
    }
  },
  "lo": {
    "rx": {
      "bytes": 24433256938,
      "packets": 55834478,
      "errs": 0,
      "drop": 0,
      "fifo": 0,
      "frame": 0,
      "compressed": 0,
      "multicast": 0
    },
    "tx": {
      "bytes": 24433256938,
      "packets": 55834478,
      "errs": 0,
      "drop": 0,
      "fifo": 0,
      "colls": 0,
      "carrier": 0,
      "compressed": 0
    }
  },
  "mgmt0": {
    "rx": {
      "bytes": 21493598205,
      "packets": 39560517,
      "errs": 0,
      "drop": 0,
      "fifo": 0,
      "frame": 0,
      "compressed": 0,
      "multicast": 0
    },
    "tx": {
      "bytes": 8893646531,
      "packets": 25626311,
      "errs": 0,
      "drop": 0,
      "fifo": 0,
      "colls": 0,
      "carrier": 0,
      "compressed": 0
    }
  },
  "mgmt1": {
    "rx": {
      "bytes": 0,
      "packets": 0,
      "errs": 0,
      "drop": 0,
      "fifo": 0,
      "frame": 0,
      "compressed": 0,
      "multicast": 0
    },
    "tx": {
      "bytes": 0,
      "packets": 0,
      "errs": 0,
      "drop": 0,
      "fifo": 0,
      "colls": 0,
      "carrier": 0,
      "compressed": 0
    }
  },
  "mgmt2": {
    "rx": {
      "bytes": 38758532460,
      "packets": 78257976,
      "errs": 0,
      "drop": 0,
      "fifo": 0,
      "frame": 0,
      "compressed": 0,
      "multicast": 0
    },
    "tx": {
      "bytes": 11666165360,
      "packets": 74856970,
      "errs": 0,
      "drop": 0,
      "fifo": 0,
      "colls": 0,
      "carrier": 0,
      "compressed": 0
    }
  },
  "usermode": {
    "rx": {
      "bytes": 449364,
      "packets": 3919,
      "errs": 0,
      "drop": 0,
      "fifo": 0,
      "frame": 0,
      "compressed": 0,
      "multicast": 0
    },
    "tx": {
      "bytes": 20174,
      "packets": 173,
      "errs": 0,
      "drop": 0,
      "fifo": 0,
      "colls": 0,
      "carrier": 0,
      "compressed": 0
    }
  },
  "veth06b8882": {
    "rx": {
      "bytes": 4148,
      "packets": 58,
      "errs": 0,
      "drop": 0,
      "fifo": 0,
      "frame": 0,
      "compressed": 0,
      "multicast": 0
    },
    "tx": {
      "bytes": 592505330,
      "packets": 664833,
      "errs": 0,
      "drop": 0,
      "fifo": 0,
      "colls": 0,
      "carrier": 0,
      "compressed": 0
    }
  },
  "veth16eb12b": {
    "rx": {
      "bytes": 4116,
      "packets": 58,
      "errs": 0,
      "drop": 0,
      "fifo": 0,
      "frame": 0,
      "compressed": 0,
      "multicast": 0
    },
    "tx": {
      "bytes": 592505130,
      "packets": 664829,
      "errs": 0,
      "drop": 0,
      "fifo": 0,
      "colls": 0,
      "carrier": 0,
      "compressed": 0
    }
  },
  "veth237e8e3": {
    "rx": {
      "bytes": 10372,
      "packets": 146,
      "errs": 0,
      "drop": 0,
      "fifo": 0,
      "frame": 0,
      "compressed": 0,
      "multicast": 0
    },
    "tx": {
      "bytes": 592517122,
      "packets": 664897,
      "errs": 0,
      "drop": 0,
      "fifo": 0,
      "colls": 0,
      "carrier": 0,
      "compressed": 0
    }
  },
  "veth2e73ef6": {
    "rx": {
      "bytes": 4148,
      "packets": 58,
      "errs": 0,
      "drop": 0,
      "fifo": 0,
      "frame": 0,
      "compressed": 0,
      "multicast": 0
    },
    "tx": {
      "bytes": 592505330,
      "packets": 664833,
      "errs": 0,
      "drop": 0,
      "fifo": 0,
      "colls": 0,
      "carrier": 0,
      "compressed": 0
    }
  },
  "veth34c4142": {
    "rx": {
      "bytes": 0,
      "packets": 0,
      "errs": 0,
      "drop": 0,
      "fifo": 0,
      "frame": 0,
      "compressed": 0,
      "multicast": 0
    },
    "tx": {
      "bytes": 592496244,
      "packets": 664758,
      "errs": 0,
      "drop": 0,
      "fifo": 0,
      "colls": 0,
      "carrier": 0,
      "compressed": 0
    }
  },
  "veth36f5f66": {
    "rx": {
      "bytes": 0,
      "packets": 0,
      "errs": 0,
      "drop": 0,
      "fifo": 0,
      "frame": 0,
      "compressed": 0,
      "multicast": 0
    },
    "tx": {
      "bytes": 592496772,
      "packets": 664770,
      "errs": 0,
      "drop": 0,
      "fifo": 0,
      "colls": 0,
      "carrier": 0,
      "compressed": 0
    }
  },
  "veth45bb9b8": {
    "rx": {
      "bytes": 3598,
      "packets": 51,
      "errs": 0,
      "drop": 0,
      "fifo": 0,
      "frame": 0,
      "compressed": 0,
      "multicast": 0
    },
    "tx": {
      "bytes": 592503148,
      "packets": 664802,
      "errs": 0,
      "drop": 0,
      "fifo": 0,
      "colls": 0,
      "carrier": 0,
      "compressed": 0
    }
  },
  "veth46b11d3": {
    "rx": {
      "bytes": 3540,
      "packets": 50,
      "errs": 0,
      "drop": 0,
      "fifo": 0,
      "frame": 0,
      "compressed": 0,
      "multicast": 0
    },
    "tx": {
      "bytes": 592503342,
      "packets": 664807,
      "errs": 0,
      "drop": 0,
      "fifo": 0,
      "colls": 0,
      "carrier": 0,
      "compressed": 0
    }
  },
  "veth4a6528b": {
    "rx": {
      "bytes": 4174,
      "packets": 59,
      "errs": 0,
      "drop": 0,
      "fifo": 0,
      "frame": 0,
      "compressed": 0,
      "multicast": 0
    },
    "tx": {
      "bytes": 592504828,
      "packets": 664822,
      "errs": 0,
      "drop": 0,
      "fifo": 0,
      "colls": 0,
      "carrier": 0,
      "compressed": 0
    }
  },
  "veth4eb3f8a": {
    "rx": {
      "bytes": 0,
      "packets": 0,
      "errs": 0,
      "drop": 0,
      "fifo": 0,
      "frame": 0,
      "compressed": 0,
      "multicast": 0
    },
    "tx": {
      "bytes": 592497024,
      "packets": 664776,
      "errs": 0,
      "drop": 0,
      "fifo": 0,
      "colls": 0,
      "carrier": 0,
      "compressed": 0
    }
  },
  "veth63d4f07": {
    "rx": {
      "bytes": 3556,
      "packets": 50,
      "errs": 0,
      "drop": 0,
      "fifo": 0,
      "frame": 0,
      "compressed": 0,
      "multicast": 0
    },
    "tx": {
      "bytes": 592503526,
      "packets": 664811,
      "errs": 0,
      "drop": 0,
      "fifo": 0,
      "colls": 0,
      "carrier": 0,
      "compressed": 0
    }
  },
  "veth65f9fa0": {
    "rx": {
      "bytes": 17214,
      "packets": 243,
      "errs": 0,
      "drop": 0,
      "fifo": 0,
      "frame": 0,
      "compressed": 0,
      "multicast": 0
    },
    "tx": {
      "bytes": 592531500,
      "packets": 665002,
      "errs": 0,
      "drop": 0,
      "fifo": 0,
      "colls": 0,
      "carrier": 0,
      "compressed": 0
    }
  },
  "veth676cb59": {
    "rx": {
      "bytes": 4116,
      "packets": 58,
      "errs": 0,
      "drop": 0,
      "fifo": 0,
      "frame": 0,
      "compressed": 0,
      "multicast": 0
    },
    "tx": {
      "bytes": 592505298,
      "packets": 664833,
      "errs": 0,
      "drop": 0,
      "fifo": 0,
      "colls": 0,
      "carrier": 0,
      "compressed": 0
    }
  },
  "veth6a20ae2": {
    "rx": {
      "bytes": 3598,
      "packets": 51,
      "errs": 0,
      "drop": 0,
      "fifo": 0,
      "frame": 0,
      "compressed": 0,
      "multicast": 0
    },
    "tx": {
      "bytes": 592503148,
      "packets": 664802,
      "errs": 0,
      "drop": 0,
      "fifo": 0,
      "colls": 0,
      "carrier": 0,
      "compressed": 0
    }
  },
  "veth72f61ea": {
    "rx": {
      "bytes": 0,
      "packets": 0,
      "errs": 0,
      "drop": 0,
      "fifo": 0,
      "frame": 0,
      "compressed": 0,
      "multicast": 0
    },
    "tx": {
      "bytes": 592497024,
      "packets": 664776,
      "errs": 0,
      "drop": 0,
      "fifo": 0,
      "colls": 0,
      "carrier": 0,
      "compressed": 0
    }
  },
  "veth7bd5de8": {
    "rx": {
      "bytes": 4116,
      "packets": 58,
      "errs": 0,
      "drop": 0,
      "fifo": 0,
      "frame": 0,
      "compressed": 0,
      "multicast": 0
    },
    "tx": {
      "bytes": 592505298,
      "packets": 664833,
      "errs": 0,
      "drop": 0,
      "fifo": 0,
      "colls": 0,
      "carrier": 0,
      "compressed": 0
    }
  },
  "veth7c0d679": {
    "rx": {
      "bytes": 4158,
      "packets": 59,
      "errs": 0,
      "drop": 0,
      "fifo": 0,
      "frame": 0,
      "compressed": 0,
      "multicast": 0
    },
    "tx": {
      "bytes": 592504770,
      "packets": 664821,
      "errs": 0,
      "drop": 0,
      "fifo": 0,
      "colls": 0,
      "carrier": 0,
      "compressed": 0
    }
  },
  "veth7ff82b2": {
    "rx": {
      "bytes": 4158,
      "packets": 59,
      "errs": 0,
      "drop": 0,
      "fifo": 0,
      "frame": 0,
      "compressed": 0,
      "multicast": 0
    },
    "tx": {
      "bytes": 592504896,
      "packets": 664824,
      "errs": 0,
      "drop": 0,
      "fifo": 0,
      "colls": 0,
      "carrier": 0,
      "compressed": 0
    }
  },
  "veth86cfdc2": {
    "rx": {
      "bytes": 3598,
      "packets": 51,
      "errs": 0,
      "drop": 0,
      "fifo": 0,
      "frame": 0,
      "compressed": 0,
      "multicast": 0
    },
    "tx": {
      "bytes": 592503232,
      "packets": 664804,
      "errs": 0,
      "drop": 0,
      "fifo": 0,
      "colls": 0,
      "carrier": 0,
      "compressed": 0
    }
  },
  "veth874a4ed": {
    "rx": {
      "bytes": 0,
      "packets": 0,
      "errs": 0,
      "drop": 0,
      "fifo": 0,
      "frame": 0,
      "compressed": 0,
      "multicast": 0
    },
    "tx": {
      "bytes": 592496370,
      "packets": 664761,
      "errs": 0,
      "drop": 0,
      "fifo": 0,
      "colls": 0,
      "carrier": 0,
      "compressed": 0
    }
  },
  "veth95ac83e": {
    "rx": {
      "bytes": 3598,
      "packets": 51,
      "errs": 0,
      "drop": 0,
      "fifo": 0,
      "frame": 0,
      "compressed": 0,
      "multicast": 0
    },
    "tx": {
      "bytes": 592503526,
      "packets": 664811,
      "errs": 0,
      "drop": 0,
      "fifo": 0,
      "colls": 0,
      "carrier": 0,
      "compressed": 0
    }
  },
  "veth9e52739": {
    "rx": {
      "bytes": 4132,
      "packets": 58,
      "errs": 0,
      "drop": 0,
      "fifo": 0,
      "frame": 0,
      "compressed": 0,
      "multicast": 0
    },
    "tx": {
      "bytes": 592505062,
      "packets": 664827,
      "errs": 0,
      "drop": 0,
      "fifo": 0,
      "colls": 0,
      "carrier": 0,
      "compressed": 0
    }
  },
  "vethb66e91f": {
    "rx": {
      "bytes": 3556,
      "packets": 50,
      "errs": 0,
      "drop": 0,
      "fifo": 0,
      "frame": 0,
      "compressed": 0,
      "multicast": 0
    },
    "tx": {
      "bytes": 592503106,
      "packets": 664801,
      "errs": 0,
      "drop": 0,
      "fifo": 0,
      "colls": 0,
      "carrier": 0,
      "compressed": 0
    }
  },
  "vethbdb5f64": {
    "rx": {
      "bytes": 4132,
      "packets": 58,
      "errs": 0,
      "drop": 0,
      "fifo": 0,
      "frame": 0,
      "compressed": 0,
      "multicast": 0
    },
    "tx": {
      "bytes": 592504870,
      "packets": 664823,
      "errs": 0,
      "drop": 0,
      "fifo": 0,
      "colls": 0,
      "carrier": 0,
      "compressed": 0
    }
  },
  "vethc28587c": {
    "rx": {
      "bytes": 5126155971,
      "packets": 15906644,
      "errs": 0,
      "drop": 0,
      "fifo": 0,
      "frame": 0,
      "compressed": 0,
      "multicast": 0
    },
    "tx": {
      "bytes": 22376563397,
      "packets": 21448851,
      "errs": 0,
      "drop": 0,
      "fifo": 0,
      "colls": 0,
      "carrier": 0,
      "compressed": 0
    }
  },
  "vethc401548": {
    "rx": {
      "bytes": 3598,
      "packets": 51,
      "errs": 0,
      "drop": 0,
      "fifo": 0,
      "frame": 0,
      "compressed": 0,
      "multicast": 0
    },
    "tx": {
      "bytes": 592503232,
      "packets": 664804,
      "errs": 0,
      "drop": 0,
      "fifo": 0,
      "colls": 0,
      "carrier": 0,
      "compressed": 0
    }
  },
  "vethc647f9c": {
    "rx": {
      "bytes": 3540,
      "packets": 50,
      "errs": 0,
      "drop": 0,
      "fifo": 0,
      "frame": 0,
      "compressed": 0,
      "multicast": 0
    },
    "tx": {
      "bytes": 592503342,
      "packets": 664807,
      "errs": 0,
      "drop": 0,
      "fifo": 0,
      "colls": 0,
      "carrier": 0,
      "compressed": 0
    }
  },
  "vethd1e8802": {
    "rx": {
      "bytes": 0,
      "packets": 0,
      "errs": 0,
      "drop": 0,
      "fifo": 0,
      "frame": 0,
      "compressed": 0,
      "multicast": 0
    },
    "tx": {
      "bytes": 592495446,
      "packets": 664739,
      "errs": 0,
      "drop": 0,
      "fifo": 0,
      "colls": 0,
      "carrier": 0,
      "compressed": 0
    }
  },
  "vethd531014": {
    "rx": {
      "bytes": 3556,
      "packets": 50,
      "errs": 0,
      "drop": 0,
      "fifo": 0,
      "frame": 0,
      "compressed": 0,
      "multicast": 0
    },
    "tx": {
      "bytes": 592503190,
      "packets": 664803,
      "errs": 0,
      "drop": 0,
      "fifo": 0,
      "colls": 0,
      "carrier": 0,
      "compressed": 0
    }
  },
  "vethe2cd54f": {
    "rx": {
      "bytes": 3582,
      "packets": 51,
      "errs": 0,
      "drop": 0,
      "fifo": 0,
      "frame": 0,
      "compressed": 0,
      "multicast": 0
    },
    "tx": {
      "bytes": 592503552,
      "packets": 664812,
      "errs": 0,
      "drop": 0,
      "fifo": 0,
      "colls": 0,
      "carrier": 0,
      "compressed": 0
    }
  },
  "vethf6ea3d7": {
    "rx": {
      "bytes": 5098678,
      "packets": 58194,
      "errs": 0,
      "drop": 0,
      "fifo": 0,
      "frame": 0,
      "compressed": 0,
      "multicast": 0
    },
    "tx": {
      "bytes": 609478062,
      "packets": 715181,
      "errs": 0,
      "drop": 0,
      "fifo": 0,
      "colls": 0,
      "carrier": 0,
      "compressed": 0
    }
  }
}
